kratos/solving_strategies/builder_and_solvers/p_multigrid/constraint_assembler.hpp

```diff
@@ -0,0 +1,270 @@
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Máté Kelemen
+//
+
+#pragma once
+
+// Project includes
+#include "solving_strategies/builder_and_solvers/p_multigrid/status_stream.hpp" // PMGStatusStream
+#include "containers/data_value_container.h" // DataValueContainer
+#include "containers/pointer_vector_set.h" // PointerVectorSet
+#include "includes/master_slave_constraint.h" // MasterSlaveConstraint
+#include "includes/indexed_object.h" // IndexedObject
+#include "includes/dof.h" // Dof
+#include "includes/variables.h" // IDENTIFIER
+#include "includes/smart_pointers.h" // KRATOS_CLASS_POINTER_DEFINITION
+
+
+namespace Kratos {
+
+
+/// @brief Enum class representing imposition methods for multifreedom constraints.
+enum class ConstraintImposition
+{
+    None                = 0,
+    MasterSlave         = 1,
+    Lagrange            = 2,
+    AugmentedLagrange   = 3,
+    Penalty             = 4
+}; // enum class ConstraintImposition
+
+
+/// @brief Interface for assembling and imposing multifreedom constraints.
+template <class TSparse, class TDense>
+class ConstraintAssembler : public DataValueContainer
+{
+public:
+    KRATOS_CLASS_POINTER_DEFINITION(ConstraintAssembler);
+
+    using DofSet = PointerVectorSet<Dof<typename TDense::DataType>>;
+
+    using ConstraintArray = PointerVectorSet<MasterSlaveConstraint, IndexedObject>;
+
+    ConstraintAssembler() noexcept
+        : ConstraintAssembler(ConstraintImposition::MasterSlave, "unnamed")
+    {}
+
+    ConstraintAssembler(ConstraintImposition Method,
+                        std::string&& rInstanceName)
+        : DataValueContainer(),
+          mRelationMatrix(),
+          mConstraintGapVector(),
+          mName(std::move(rInstanceName))
+    {
+        std::string method_name;
+
+        switch (Method) {
+            case ConstraintImposition::None:
+                method_name = "none";
+                break;
+            case ConstraintImposition::MasterSlave:
+                method_name = "master_slave";
+                break;
+            case ConstraintImposition::AugmentedLagrange:
+                method_name = "augmented_lagrange";
+                break;
+            default:
+                // Other imposition methods are not implemented yet.
+                KRATOS_ERROR << "Unsupported constraint imposition: " << (int)Method;
+        } // switch Method
+
+        this->SetValue(ConstraintAssembler::GetImpositionVariable(), method_name);
+    }
+
+    /// @details Define an overriding virtual destructor to ensure compile time errors
+    //           if the base class' destructor turns non-virtual.
+    virtual ~ConstraintAssembler() override = default;
+
+    /// @brief Allocate memory for the constraint gap vector and relation matrix, and compute its topology.
+    /// @details This function is responsible for large memory allocations, as well as computing
```

Comment:
My comment about the comments sill there

Reply:
done